This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [7]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = ''

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [8]:
# First, import the relevant modules
import requests

In [11]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
r = requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2018-10-30&end_date=2018-10-30&api_key=API_KEY')

In [12]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
print(r.json())

{'dataset': {'id': 10095370, 'dataset_code': 'AFX_X', 'database_code': 'FSE', 'name': 'Carl Zeiss Meditec (AFX_X)', 'description': 'Stock Prices for Carl Zeiss Meditec (AFX) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', 'refreshed_at': '2018-10-31T22:30:31.448Z', 'newest_available_date': '2018-10-31', 'oldest_available_date': '2000-06-07', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': None, 'transform': None, 'column_index': None, 'start_date': '2018-10-30', 'end_date': '2018-10-30', 'data': [['2018-10-30', None, 72.5, 70.85, 72.05, None, 65279.0, 4687185.0, None, None, None]], 'collapse': None, 'order': None, 'database_id': 6129}}


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [18]:
#1. Collect data from FSE, for ticker AFX_X for whole year 2017
req2017 = requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=c-KftEJKUMaM_4vYbHRk')

In [21]:
#2. Convert JSON object into a Python dictionary
import json
stock_price_data = req2017.json()

In [27]:
#3. Calculate highest and lowest opening prices in 2017
opening_prices = []
for data in stock_price_data['dataset']['data']:
    if data[1] is not None:
        opening_prices.append(data[1])
highest = max(opening_prices)
lowest = min(opening_prices)
print(highest, lowest)

53.11 34.0


The highest opening price for the stock in 2017 was 53.11 Euros and the lowest opening price for the stock that year was 34 Euros.

In [41]:
#4. Calculate largest change in one day based on high and low price
high_prices = []
low_prices = []
for data in stock_price_data['dataset']['data']:
    high_prices.append(data[2])
    low_prices.append(data[3])
change = []
for x, y in zip(high_prices, low_prices):
    if (x is None) or (y is None):
        change.append(None)
    else:
        change.append(x - y)
print(max([val for val in change if val is not None]))

2.8100000000000023


The largest change in the stock price in any one day based on high and low price was about 2.81 Euros.

In [42]:
#5. Calculate largest change between any two days based on closing price
closing_prices = []
for data in stock_price_data['dataset']['data']:
    closing_prices.append(data[4])
print(max(closing_prices) - min(closing_prices))

19.03


The largest change in the stock price between any two days, based on closing price, was 19.03 Euros.

In [47]:
#6. Calculate average daily trading volume
daily_trading_volumes = []
for data in stock_price_data['dataset']['data']:
    daily_trading_volumes.append(data[6])
print(sum(daily_trading_volumes) / len(daily_trading_volumes))

89124.33725490196


The average daily trading volume of shares of Carl Zeiss Meditec in 2017 was about 89,124 shares.